snow/engine/common/engine.go

```diff
@@ -389,15 +406,13 @@ type AppResponseHandler interface {
 		response []byte,
 	) error
 
-	// Notify this engine that an AppRequest it issued has failed.
-	//
-	// This function will be called if an AppRequest message with nodeID and
-	// requestID was previously sent by this engine and will not receive a
-	// response.
+	// AppRequestFailed is called when a pending AppRequest failed due to either
+	// a timeout or an application-defined reason.
 	AppRequestFailed(
 		ctx context.Context,
 		nodeID ids.NodeID,
 		requestID uint32,
+		err error,
```

Comment:
Can we document what it means if `err == nil`? Is that possible?

Reply:
Hmm `err` should never be `nil` because this function is only invoked if an error condition occurs... so I feel like we should guarantee non-nil here. I think it's fine if the implementation panics on a `nil` error since that's very unexpected. I can document that `err` is always non-nil.

Reply:
OK, there's also a place in `TestRouterDropMessage` we pass a `nil` error so maybe we should remove that as well?

Reply:
Yeah that should be removed.